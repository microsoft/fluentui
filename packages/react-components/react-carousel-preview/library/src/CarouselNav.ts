export * from './components/CarouselNav/index';
