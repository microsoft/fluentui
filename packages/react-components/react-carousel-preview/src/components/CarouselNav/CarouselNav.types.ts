import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type CarouselNavSlots = {
  root: Slot<'div'>;
};

/**
 * CarouselNav Props
 */
export type CarouselNavProps = ComponentProps<CarouselNavSlots> & {};

/**
 * State used in rendering CarouselNav
 */
export type CarouselNavState = ComponentState<CarouselNavSlots>;
// TODO: Remove semicolon from previous line, uncomment next line, and provide union of props to pick from CarouselNavProps.
// & Required<Pick<CarouselNavProps, 'propName'>>
