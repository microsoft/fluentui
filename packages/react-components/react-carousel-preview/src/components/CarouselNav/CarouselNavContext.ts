import * as React from 'react';

const carouselNavContext = React.createContext<string | undefined>(undefined);

export const carouselNavContextDefaultValue = '';

export const useCarouselNavContext = () => React.useContext(carouselNavContext) ?? carouselNavContextDefaultValue;

export const CarouselNavContextProvider = carouselNavContext.Provider;
