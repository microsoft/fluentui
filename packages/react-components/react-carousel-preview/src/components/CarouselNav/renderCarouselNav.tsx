/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type { CarouselNavState, CarouselNavSlots } from './CarouselNav.types';

/**
 * Render the final JSX of CarouselNav
 */
export const renderCarouselNav_unstable = (state: CarouselNavState) => {
  assertSlots<CarouselNavSlots>(state);

  // TODO Add additional slots in the appropriate place
  return <state.root />;
};
