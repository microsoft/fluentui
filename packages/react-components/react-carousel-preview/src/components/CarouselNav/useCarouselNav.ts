import * as React from 'react';
import { getIntrinsicElementProps, slot } from '@fluentui/react-utilities';
import type { CarouselNavProps, CarouselNavState } from './CarouselNav.types';

/**
 * Create the state required to render CarouselNav.
 *
 * The returned state can be modified with hooks such as useCarouselNavStyles_unstable,
 * before being passed to renderCarouselNav_unstable.
 *
 * @param props - props from this instance of CarouselNav
 * @param ref - reference to root HTMLDivElement of CarouselNav
 */
export const useCarouselNav_unstable = (props: CarouselNavProps, ref: React.Ref<HTMLDivElement>): CarouselNavState => {
  return {
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: slot.always(
      getIntrinsicElementProps('div', {
        ref,
        ...props,
      }),
      { elementType: 'div' },
    ),
  };
};
