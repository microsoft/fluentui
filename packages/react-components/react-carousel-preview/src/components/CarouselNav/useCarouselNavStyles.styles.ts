import { makeStyles, mergeClasses } from '@griffel/react';
import type { SlotClassNames } from '@fluentui/react-utilities';
import type { CarouselNavSlots, CarouselNavState } from './CarouselNav.types';

export const carouselNavClassNames: SlotClassNames<CarouselNavSlots> = {
  root: 'fui-CarouselNav',
  // TODO: add class names for all slots on CarouselNavSlots.
  // Should be of the form `<slotName>: 'fui-CarouselNav__<slotName>`
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    // TODO Add default styles for the root element
  },

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the CarouselNav slots based on the state
 */
export const useCarouselNavStyles_unstable = (state: CarouselNavState): CarouselNavState => {
  const styles = useStyles();
  state.root.className = mergeClasses(carouselNavClassNames.root, styles.root, state.root.className);

  // TODO Add class names to slots, for example:
  // state.mySlot.className = mergeClasses(styles.mySlot, state.mySlot.className);

  return state;
};
