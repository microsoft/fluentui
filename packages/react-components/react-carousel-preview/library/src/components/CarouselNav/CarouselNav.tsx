import type { ForwardRefComponent } from '@fluentui/react-utilities';
import * as React from 'react';

import type { CarouselNavProps } from './CarouselNav.types';
import { useCarouselNavContextValues_unstable } from './CarouselNavContext';
import { renderCarouselNav_unstable } from './renderCarouselNav';
import { useCarouselNav_unstable } from './useCarouselNav';
import { useCarouselNavStyles_unstable } from './useCarouselNavStyles.styles';

/**
 * Used to jump to a card based on index, using arrow navigation via Tabster.
 *
 * The children of this component will be wrapped in a context to
 * provide the appropriate value based on their index position.
 */
export const CarouselNav: ForwardRefComponent<CarouselNavProps> = React.forwardRef((props, ref) => {
  const state = useCarouselNav_unstable(props, ref);
  const contextValues = useCarouselNavContextValues_unstable(state);

  useCarouselNavStyles_unstable(state);
  // TODO update types in packages/react-components/react-shared-contexts/src/CustomStyleHooksContext/CustomStyleHooksContext.ts
  // https://github.com/microsoft/fluentui/blob/master/rfcs/react-components/convergence/custom-styling.md
  // useCustomStyleHook_unstable('useCarouselNavStyles_unstable')(state);

  return renderCarouselNav_unstable(state, contextValues);
});

CarouselNav.displayName = 'CarouselNav';
