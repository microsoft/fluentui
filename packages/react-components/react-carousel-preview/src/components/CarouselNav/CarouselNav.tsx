import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useCarouselNav_unstable } from './useCarouselNav';
import { renderCarouselNav_unstable } from './renderCarouselNav';
import { useCarouselNavStyles_unstable } from './useCarouselNavStyles.styles';
import type { CarouselNavProps } from './CarouselNav.types';

/**
 * Used to jump to a card based on index, using arrow navigation via Tabster.
 *
 * The children of this component will be wrapped in a context to
 * provide the appropriate value based on their index position.
 */
export const CarouselNav: ForwardRefComponent<CarouselNavProps> = React.forwardRef((props, ref) => {
  const state = useCarouselNav_unstable(props, ref);

  useCarouselNavStyles_unstable(state);
  // TODO update types in packages/react-components/react-shared-contexts/src/CustomStyleHooksContext/CustomStyleHooksContext.ts
  // https://github.com/microsoft/fluentui/blob/master/rfcs/react-components/convergence/custom-styling.md
  // useCustomStyleHook_unstable('useCarouselNavStyles_unstable')(state);
  return renderCarouselNav_unstable(state);
});

CarouselNav.displayName = 'CarouselNav';
