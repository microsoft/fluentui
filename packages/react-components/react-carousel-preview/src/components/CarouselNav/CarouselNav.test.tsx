import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { CarouselNav } from './CarouselNav';

describe('CarouselNav', () => {
  isConformant({
    Component: CarouselNav,
    displayName: 'CarouselNav',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<CarouselNav>Default CarouselNav</CarouselNav>);
    expect(result.container).toMatchSnapshot();
  });
});
