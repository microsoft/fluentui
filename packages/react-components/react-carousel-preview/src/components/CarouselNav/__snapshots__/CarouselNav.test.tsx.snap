// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselNav renders a default state 1`] = `
<div>
  <div
    class="fui-CarouselNav"
  >
    Default CarouselNav
  </div>
</div>
`;
